py-polars/tests/unit/interop/test_interop.py

```diff
@@ -749,3 +749,32 @@ def test_compat_level(monkeypatch: pytest.MonkeyPatch) -> None:
     assert len(df.write_ipc_stream(None).getbuffer()) == 544
     assert len(df.write_ipc_stream(None, compat_level=oldest).getbuffer()) == 672
     assert len(df.write_ipc_stream(None, compat_level=newest).getbuffer()) == 544
+
+
+def test_df_pycapsule_interface() -> None:
+    class PyCapsuleStreamHolder:
+        """
+        Hold the Arrow C Stream pycapsule.
+
+        A class that exposes _only_ the Arrow C Stream interface via Arrow PyCapsules.
+        This ensures that pyarrow is seeing _only_ the `__arrow_c_stream__` dunder, and
+        that nothing else (e.g. the dataframe or array interface) is actually being
+        used.
+        """
+
+        capsule: object
+
+        def __init__(self, capsule: object) -> None:
+            self.capsule = capsule
+
+        def __arrow_c_stream__(self, requested_schema: object) -> object:
+            return self.capsule
+
+    df = pl.DataFrame({"a": [1, 2, 3], "b": ["a", "b", "c"]})
+    out = pa.table(PyCapsuleStreamHolder(df.__arrow_c_stream__(None)))
+    assert df.shape == out.shape
+    assert df.schema.names() == out.schema.names
```

Comment:
You could drop `df` just now and make sure that the recreated `df2` below still gets the expected contents (instead of crashing or whatever else).

Reply:
I updated the test to not hold a bare capsule, but rather call the underlying object's `__arrow_c_stream__` method. I'm not sure what you're suggesting this test, since I need to check below that `df` and `df2` are equal. Are you suggesting after that I should drop `df` again? That isn't possible when this utility class doesn't hold bare capsules